payment/app.py

```diff
@@ -1,87 +1,113 @@
-                              password=os.environ['REDIS_PASSWORD'],
-                              db=int(os.environ['REDIS_DB']))
-
-
-def close_db_connection():
-    db.close()
 
+username = os.environ['MONGODB_USERNAME']
+password = os.environ['MONGODB_PASSWORD']
+hostname = os.environ['MONGODB_HOSTNAME']
+database = os.environ['MONGODB_DATABASE']
+gateway_url = os.environ['GATEWAY_URL']
 
-atexit.register(close_db_connection)
+app.config["MONGO_URI"] = f"mongodb://{username}:{password}@{hostname}:27017/{database}"
 
+mongo = PyMongo(app)
+db = mongo.db
+user_collection = db.users
+paid_order_colleciton = db.paid_orders
 
 @app.post('/create_user')
 def create_user():
-    user_id = db.incr('user_id')
-    user = {'user_id': user_id, 'credit': 0}
-    db.hmset(f'user:{user_id}', user)
-    return {"user_id": user_id}, 200
+    user = user_collection.insert_one({'credit': 0}) 
+    user_id = user.inserted_id
+    return {'user_id': str(user_id)}
+
 
 
 @app.get('/find_user/<user_id>')
 def find_user(user_id: str):
-    user = db.hmget(f'user:{user_id}', 'credit')
+    user = user_collection.find_one({'_id': ObjectId(user_id)})
     if None in user:
         return {'Error': 'User not found'}, 404
-    return {"user_id": int(user_id), "credit": int(user[0])}, 200
+    return {'user_id': user_id, 'credit': user['credit']}, 200
 
 
 @app.post('/add_funds/<user_id>/<amount>')
 def add_credit(user_id: str, amount: int):
     if int(amount) < 0:
         return {'Error': 'Amount must be positive'}, 400
-    if not db.exists(f'user:{user_id}'):
-        return {'Error': 'User not found'}, 400
-    db.hincrby(f'user:{user_id}', 'credit', int(amount))
-    return 'Success', 200
+    user = user_collection.find_one_or_404({'_id': ObjectId(user_id)})
+    credit: int = int(user['credit'])
+    newCredit = credit + int(amount)
+    result = user_collection.update_one(
+        # Filter to check if the credit is unchanged
+        {'_id': ObjectId(user_id), 'credit': { '$eq': credit}},
+        {'$set': {'credit': newCredit}}
+    )
+
+    if result.modified_count != 1:
+        return {'Error': 'The credit was updated by someone else'}, 400
+    
+    return {'Success': 'Credit is updated successfully'}, 200
+
 
 
 @app.post('/pay/<user_id>/<order_id>/<amount>')
 def remove_credit(user_id: str, order_id: str, amount: int):
-    credit = db.hget(f'user:{user_id}', 'credit')
-    if int(credit) < int(amount):
-        return 'Not enough credit', 400
-    order = db.exists(f'paid_orders:{order_id}')
+    
+    order = paid_order_colleciton.find_one({'order_id': order_id})
 
     if order:
         return 'Order already paid', 401
 
-    p = db.pipeline(transaction=True)
-    p.hincrby(f'user:{user_id}', 'credit', -int(amount))
-    p.hmset(f'paid_orders:{order_id}', {
-            'amount_paid': amount, 'user_id': user_id})
-    p.execute()
+    user = user_collection.find_one_or_404({'_id': ObjectId(user_id)})
```

Comment:
What does this do if the user cannot be found?

Reply:
![image](https://github.com/pascalbenschopTU/WDM_project/assets/45072611/81885919-039d-4d29-ad71-0e3c5dab9d9c)

It should return a 404 on the incoming request

Reply:
@pascalbenschopTU I did some testing with it and it returns a generic 404 page to the user and stops executing the endpoint. However, in hindsight it might be more useful, and readable, if we return a custom error. I have updated the 404 calls to a normal find one call with a None check